src/impl/tpls/KokkosBlas3_gemm_tpl_spec_decl.hpp

```diff
@@ -335,6 +335,127 @@ KOKKOSBLAS3_CGEMM_BLAS( Kokkos::LayoutRight, Kokkos::LayoutRight, Kokkos::Layout
+  size_A chunkSize;      // the local length of each team's share on the dot product  
+  
+
+  DotBasedGEMM(const scalar_A& alpha_, const AV& A_, const BV& B_, const scalar_C& beta_, const CV& C_):A(A_),B(B_),C(C_),alpha(alpha_),beta(beta_) 
+  {
+    numCrows = C.extent(0);
+    numCcols = C.extent(1);
+    dotSize = A.extent(0);
+  }
+
+  void run() {
+
+    size_C ndots = C.extent(0) * C.extent(1);       // Number of dot products
+    size_C appxNumTeams = (dotSize * ndots) / 4096; // First, estimate the appxNumTeams
+                                                    // each team performs 4096 mult-add ops
+
+    // Adjust appxNumTeams in case it is too small or too large
+    if(appxNumTeams < 32)
+      appxNumTeams = 32;
+    if(appxNumTeams > 1024)
+      appxNumTeams = 1024;
+
+    // If there are more dot products than the number of teams,
+    // then set the number of teams to be number of dot products
+    // and each team will perform only one dot product.
+    // We don't want a team to perform more than one dot product.
+    if(ndots >= appxNumTeams) {
+      numTeams = ndots;
+      numDivPerDot = 1;
+    }
+    // If there are more teams than dot products, each dot product can
+    // potentially be performed by multiple teams. First, compute 
+    // numDivPerDot as an integer (take the floor, not ceiling), then,
+    // compute actual number of teams by using this factor.
+    else{
+      numDivPerDot = appxNumTeams / ndots;
+      numTeams = ndots * numDivPerDot;
+    }
+
+    // Determine the local length for the dot product
+    chunkSize = dotSize / numDivPerDot;
+    if(numDivPerDot > 1)
+      chunkSize++;
+
+    // Initialize C matrix as beta*C
+    Kokkos::RangePolicy<TagInit, ExecSpace> policy1(0, ndots);
+    Kokkos::parallel_for("Initialize C for Dot Product Based GEMM", policy1, *this);
+
+    // Multiply alpha*A^TB and add it to beta*C
+    Kokkos::TeamPolicy<TagMult, ExecSpace> policy2(numTeams, Kokkos::AUTO);
+    Kokkos::parallel_for("Perform Dot Product Based GEMM", policy2, *this);
+
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  void operator() (const TagInit&, const size_C &i) const {
+    const size_C rowId = i / numCcols;
+    const size_C colId = i % numCcols;
+    C(rowId, colId) = beta * C(rowId, colId);
+  }
+
+
+  KOKKOS_INLINE_FUNCTION
+  void operator() (const TagMult&, const typename Kokkos::TeamPolicy<>::member_type& teamMember) const {
+
+    const size_C globalRank = teamMember.league_rank();
+    const size_C localRank = globalRank % numDivPerDot;
+    const size_C i = globalRank / numDivPerDot;
+    const size_C rowId = i / numCcols;
+    const size_C colId = i % numCcols;
+    
+
+    scalar_C result = 0;
+    const size_A baseInd = chunkSize*localRank; 
+    Kokkos::parallel_reduce( Kokkos::TeamThreadRange(teamMember, chunkSize), [&]( const size_A k, scalar_C &update ) {
+	if(baseInd + k < dotSize)
+	  update += alpha * A(baseInd+k, rowId) * B(baseInd+k, colId);
+      }, result );
+
```

Comment:
Does this give a correct result ? Do you have a test for this ? In my opinion, you might want to use TeamVectorRange instead of TeamThreadRange. If you use TeamThreadRange (mapping to threadIdx.y), there are extra threads (threadIdx.x that is usually mapped to ThreadVectorRange) idling or doing something weird. As you use AUTO setting, you do not guarantee the vector size = 1.

Reply:
Yes, it gives correct results for the cases I tested it. I manually tested it by comparing the resulting matrix against that of CUBLAS' gemm for different sizes of M,N,K. The unit tests pass;  e.g., https://github.com/kokkos/kokkos-kernels/blob/d86db111124cea12e23dd3447b6c307f96ef7439/unit_test/blas/Test_Blas3_gemm.hpp#L203 tests it.

I will look into how I can use TeamVectorRange instead of TeamThreadRange. Thanks @kyungjoo-kim !

Reply:
@kyungjoo-kim I couldn't find any explained examples of using TeamVectorRange, so I just replaced TeamThreadRange with TeamVectorRange :) It works and gives correct results. However, timing didn't change with this..

Reply:
Let's discuss this for the next PR. Some people are waiting for this PR.